// @serverless-stack/core/dist/cli/StacksBuilder.js
import chokidar from "chokidar";
import { interpret, actions, assign, createMachine } from "xstate";
import { Stacks } from "../stacks/index.js";
import { synth } from "../index.js";
import path from "path";
import fs from "fs-extra";
import crypto from "crypto";
function stub(name, duration = 1000) {
    return function () {
        console.log(name);
        return new Promise(r => setTimeout(r, duration));
    };
}
const machine = createMachine({
    initial: "idle",
    id: "stacksBuilder",
    states: {
        idle: {
            initial: "none",
            on: {
                FILE_CHANGE: "building"
            },
            states: {
                none: {},
                unchanged: {},
                deployed: {}
            }
        },
        failed: {
            on: {
                FILE_CHANGE: "building"
            },
            states: {
                build: {},
                synth: {},
                deploy: {}
            }
        },
        building: {
            entry: assign({
                dirty: () => false
            }),
            invoke: {
                src: "build",
                onDone: [
                    {
                        cond: "isDirty",
                        target: "building"
                    },
                    {
                        target: "synthing"
                    }
                ],
                onError: [
                    {
                        cond: "isDirty",
                        target: "building"
                    },
                    {
                        target: "failed.build"
                    }
                ]
            }
        },
        synthing: {
            invoke: {
                src: "synth",
                onDone: [
                    {
                        cond: "isDirty",
                        target: "building"
                    },
                    {
                        cond: "isChanged",
                        target: "deployable",
                        actions: actions.assign({
                            pendingHash: (_, evt) => evt.data
                        })
                    },
                    {
                        target: "idle.unchanged"
                    }
                ],
                onError: [
                    {
                        cond: "isDirty",
                        target: "building"
                    },
                    {
                        target: "failed.synth"
                    }
                ]
            }
        },
        deployable: {
            on: {
                TRIGGER_DEPLOY: "deploying",
                FILE_CHANGE: "building"
            }
        },
        deploying: {
            invoke: {
                src: "deploy",
                onDone: [
                    {
                        cond: "isDirty",
                        target: "building",
                        actions: actions.assign({
                            deployedHash: ctx => ctx.pendingHash
                        })
                    },
                    {
                        target: "idle.deployed",
                        actions: actions.assign({
                            deployedHash: ctx => ctx.pendingHash
                        })
                    }
                ],
                onError: [
                    {
                        cond: "isDirty",
                        target: "building"
                    },
                    {
                        target: "failed.deploy"
                    }
                ]
            }
        }
    },
    on: {
        FILE_CHANGE: {
            actions: actions.assign({
                dirty: _ctx => true
            })
        }
    }
}, {
    services: {
        build: stub("build"),
        deploy: stub("deploy"),
        synth: stub("synth")
    },
    guards: {
        isDirty,
        isChanged
    }
});
// TODO: The arguments here are hacky because we need to access code from cdkHelper. Should be refactored so that cdkHelpers don't really exist and everything is done inside here.
export function useStacksBuilder(root, bus, config, cdkOptions, deployFunc, initialStacksData) {
    async function publishStacksDeployed(stacksData) {
        const metadata = await Stacks.metadata(root, config);
        bus.publish("stacks.deployed", {
            metadata,
            stacksData,
        });
    }
    const cdkOutPath = path.join(root, cdkOptions.output);
    const service = interpret(machine
        .withConfig({
        services: {
            build: async () => {
                await Stacks.build(root, config);
                if (config.main.endsWith(".js"))
                    setTimeout(() => {
                        const result = Stacks.check(root, config);
                        console.log(Stacks.formatDiagnostics(result).join("\n"));
                    }, 1);
            },
            synth: async () => {
                await synth(cdkOptions);
                return generateChecksum(cdkOutPath);
            },
            deploy: async () => {
                const stacksData = await deployFunc(cdkOptions);
                publishStacksDeployed(stacksData);
            }
        }
    })
        .withContext({
        dirty: false,
        pendingHash: "",
        deployedHash: generateChecksum(cdkOutPath)
    }));
    chokidar
        .watch(path.dirname(config.main) + "/**/*", {
        persistent: true,
        ignoreInitial: true,
        followSymlinks: false,
        ignored: ["**/node_modules/**", "**/.build/**", "**/.sst/**"]
    })
        .on("change", () => {
        service.send("FILE_CHANGE");
    });
    publishStacksDeployed(initialStacksData);
    service.start();
    return service;
}
function isChanged(ctx, evt) {
    return evt.data !== ctx.deployedHash;
}
function isDirty(ctx) {
    return ctx.dirty;
}
function generateChecksum(cdkOutPath) {
    const manifestPath = path.join(cdkOutPath, "manifest.json");
    const cdkManifest = fs.readJsonSync(manifestPath);
    const checksumData = Object.keys(cdkManifest.artifacts)
        .filter((key) => cdkManifest.artifacts[key].type === "aws:cloudformation:stack")
        .map((key) => {
        const { templateFile } = cdkManifest.artifacts[key].properties;
        const templatePath = path.join(cdkOutPath, templateFile);
        const templateContent = fs.readFileSync(templatePath);
        return templateContent;
    })
        .join("\n");
    const hash = crypto
        .createHash("sha256")
        .update(checksumData)
        .digest("hex");
    return hash;
}
