// @serverless-stack/core/dist/aws-sdk/sdk_ini_file.js
/**
 * A reimplementation of JS AWS SDK's SharedIniFile class
 *
 * We need that class to parse the ~/.aws/config file to determine the correct
 * region at runtime, but unfortunately it is private upstream.
 */
import * as os from "os";
import * as path from "path";
import * as AWS from "aws-sdk";
import fs from "fs-extra";
export class SharedIniFile {
    constructor(options) {
        options = options || {};
        this.isConfig = options.isConfig === true;
        this.filename = options.filename || this.getDefaultFilepath();
    }
    async getProfile(profile) {
        await this.ensureFileLoaded();
        const profileIndex = profile !== AWS.util.defaultProfile && this.isConfig
            ? "profile " + profile
            : profile;
        return this.parsedContents[profileIndex];
    }
    getDefaultFilepath() {
        return path.join(os.homedir(), ".aws", this.isConfig ? "config" : "credentials");
    }
    async ensureFileLoaded() {
        if (this.parsedContents) {
            return;
        }
        if (!(await fs.pathExists(this.filename))) {
            this.parsedContents = {};
            return;
        }
        const contents = (await fs.readFile(this.filename)).toString();
        this.parsedContents = AWS.util.ini.parse(contents);
    }
}
