// @serverless-stack/core/dist/telemetry/environment.js
import os from "os";
import path from "path";
import fs from "fs-extra";
import ciInfo from "ci-info";
let data;
export function getEnvironmentData() {
    if (data) {
        return data;
    }
    const cpus = os.cpus() || [];
    const isSeed = !!process.env.SEED_APP_NAME;
    data = {
        // Software information
        systemPlatform: os.platform(),
        systemRelease: os.release(),
        systemArchitecture: os.arch(),
        // Machine information
        cpuCount: cpus.length,
        cpuModel: cpus.length ? cpus[0].model : null,
        cpuSpeed: cpus.length ? cpus[0].speed : null,
        memoryInMb: Math.trunc(os.totalmem() / Math.pow(1024, 2)),
        // Environment information
        isCI: ciInfo.isCI || isSeed,
        ciName: (isSeed ? "Seed" : ciInfo.name) || null,
        sstVersion: getSSTVersion(),
    };
    return data;
}
import { createRequire } from "module";
const require = createRequire(import.meta.url);
function getSSTVersion() {
    const module = path.join(require.resolve("@serverless-stack/core"), "../../", "package.json");
    const packageJson = fs.readJsonSync(module);
    return packageJson.version;
}
